doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,422 @@
+
+- It will be the responsibility of the `DurableCatalogState` to commit the `Transaction` instead of
+  the `Transaction` committing itself.
+- `Transaction` will not hold onto a mutable reference to a `Stash`. Though it will need a mutable
+  reference to some part of `DurableCatalogState` to prevent multiple concurrent transactions.
+- To avoid cloning the entire catalog state in the persist implementation, we'll
+  update `TableTransaction`s to use `CoW`s.
+
+### Catalog Debug Tool
+
+The catalog debug tool is modeled after the stash debug tool and allows users to inspect and modify
+the durable catalog state. It will provide the following functionality (descriptions are mostly
+taken from the existing stash debug tool):
+
+- Dump: Dumps the catalog contents in a human-readable format. Includes JSON for each key and value
+  that can be hand edited and then passed to the `edit` or `delete` commands.
+    - [optional arg] `target`: Write output to specified path. Default to stdout.
+- Edit: Edits a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+    - [arg] `value`: JSON encoded value of object.
+- Delete: Deletes a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+- Upgrade Check: Checks if the specified catalog could be upgraded from its state to the catalog at
+  the version of this binary. Prints a success message or error message. Exits with 0 if the upgrade
+  would succeed, otherwise non-zero. Can be used on a running `environmentd`. Operates without
+  interfering with it or committing any data to that catalog.
+    - [optional arg] `cluster-replica-sizes`: Valid cluster replica sizes.
+
+The tool will have two modes:
+
+- Stash: Connects to a stash.
+- Persist: Connects to persist.
+
+The tool will initialize a `Box<dyn DurableCatalogState>` and use the trait methods to implement
+each functionality. `dyn DurableCatalogState` is safe because the trait is designed to be object
+safe.
+
+- Dump: Will use the `dump` method.
+- Edit and Delete: Will be implemented using a catalog transaction.
+- Upgrade Check: Will use the `check_open` method.
+
+NB: There is an open issue that you need two different versions of the stash debug tool to perform
+upgrade checks: https://github.com/MaterializeInc/materialize/issues/21121. This proposal will not
+fix that issue and suffers from the exact same issue.
+
+### Migrations
+
+This section is currently hand-wavy and needs the most additional design work.
+
+Stash migration ownership is currently tangled between the adapter crate and the stash crate,
+see [stash: Untangle the stash config collection](https://github.com/MaterializeInc/materialize/issues/20922).
+Ideally the adapter would be in charge of how and why to migrate the catalog contents while the
+stash provides the primitives needed to migrate a stash collection.
```

Comment:
+1 this sounds great to me. Retaining some variant of the [migration helpers](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/upgrade.rs#L68) in the `mz_stash` crate, but moving the logic for version and the framework into the `mz_adapter` crate (or a new crate if we create one for this) would work well.